File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/websocket/ToriiWebSocketClient.java
package org.hyperledger.iroha.android.client.websocket;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import org.hyperledger.iroha.android.client.ClientObserver;
import org.hyperledger.iroha.android.client.ClientResponse;
import org.hyperledger.iroha.android.client.transport.TransportRequest;
import org.hyperledger.iroha.android.client.transport.TransportWebSocket;

/**
 * WebSocket client built on the transport abstractions shared between JVM and Android targets.
 *
 * <p>By default the platform connector is used (OkHttp on Android, JDK connector on JVM). Android
 * callers can still inject an explicit OkHttp connector to reuse a shared client instance.
 */
public final class ToriiWebSocketClient {

  /** Connector hook so platforms can supply their preferred WebSocket implementation. */
  @FunctionalInterface
  public interface WebSocketConnector {
    CompletableFuture<TransportWebSocket> connect(
        TransportRequest request,
        ToriiWebSocketOptions options,
        TransportWebSocket.Listener listener,
        Map<String, String> defaultHeaders);
  }

  private final URI baseUri;
  private final Map<String, String> defaultHeaders;
  private final List<ClientObserver> observers;
  private final WebSocketConnector connector;

  private ToriiWebSocketClient(final Builder builder) {
    this.baseUri = builder.baseUri;
    this.defaultHeaders = Map.copyOf(builder.defaultHeaders);
    this.observers = List.copyOf(builder.observers);
    this.connector =
        builder.connector != null
            ? builder.connector
            : PlatformWebSocketConnector.createDefault();
  }

  public static Builder builder() {
    return new Builder();
  }

  /** Opens a WebSocket session against the given path. */
  public ToriiWebSocketSession connect(
      final String path,
      final ToriiWebSocketOptions options,
      final ToriiWebSocketListener listener) {
    Objects.requireNonNull(listener, "listener");
    final ToriiWebSocketOptions resolved =
        options != null ? options : ToriiWebSocketOptions.defaultOptions();
    final TransportRequest request = buildRequest(path, resolved);
    notifyRequest(request);
    final ToriiWebSocketSessionImpl session = new ToriiWebSocketSessionImpl();
    final TransportWebSocket.Listener adapter = new Adapter(listener, session);
    connector
        .connect(request, resolved, adapter, defaultHeaders)
        .whenComplete(
            (socket, throwable) -> {
              if (throwable != null) {
                session.fail(throwable);
                notifyFailure(request, throwable);
                listener.onError(session, throwable);
                return;
              }
              session.attach(socket);
              notifyResponse(request);
            });
    return session;
  }

  private TransportRequest buildRequest(final String path, final ToriiWebSocketOptions options) {
    final URI httpUri = appendQueryParameters(resolvePath(path), options.queryParameters());
    final URI wsUri = toWebSocketUri(httpUri);
    final TransportRequest.Builder builder =
        TransportRequest.builder()
            .setMethod("GET")
            .setUri(wsUri)
            .setTimeout(options.connectTimeout());
    defaultHeaders.forEach(builder::addHeader);
    options.headers().forEach(builder::addHeader);
    if (!options.subprotocols().isEmpty()) {
      builder.addHeader("Sec-WebSocket-Protocol", String.join(",", options.subprotocols()));
    }
    return builder.build();
  }

  private URI resolvePath(final String path) {
    if (path == null || path.isBlank()) {
      return baseUri;
    }
    if (path.startsWith("http://") || path.startsWith("https://") || path.startsWith("ws://") || path.startsWith("wss://")) {
      return URI.create(path);
    }
    final String normalized = path.startsWith("/") ? path.substring(1) : path;
    final String base = baseUri.toString();
    final String joined = base.endsWith("/") ? base + normalized : base + "/" + normalized;
    return URI.create(joined);
  }

  private static URI appendQueryParameters(final URI target, final Map<String, String> params) {
    if (params.isEmpty()) {
      return target;
    }
    final StringBuilder builder = new StringBuilder(target.toString());
    final String query = encodeQuery(params);
    if (target.getQuery() == null || target.getQuery().isEmpty()) {
      builder.append(target.toString().contains("?") ? "&" : "?");
    } else {
      builder.append("&");
    }
    builder.append(query);
    return URI.create(builder.toString());
  }

  private static String encodeQuery(final Map<String, String> params) {
    final StringBuilder builder = new StringBuilder();
    boolean first = true;
    for (final Map.Entry<String, String> entry : params.entrySet()) {
      if (!first) {
        builder.append('&');
      }
      builder
          .append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8))
          .append('=')
          .append(URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8));
      first = false;
    }
    return builder.toString();
  }

  private static URI toWebSocketUri(final URI httpUri) {
    final String scheme =
        "https".equalsIgnoreCase(httpUri.getScheme()) || "wss".equalsIgnoreCase(httpUri.getScheme())
            ? "wss"
            : "ws";
    return URI.create(
        scheme
            + "://"
            + httpUri.getAuthority()
            + (httpUri.getRawPath() == null ? "" : httpUri.getRawPath())
            + (httpUri.getRawQuery() == null ? "" : "?" + httpUri.getRawQuery()));
  }

  private void notifyRequest(final TransportRequest request) {
    for (final ClientObserver observer : observers) {
      observer.onRequest(request);
    }
  }

  private void notifyResponse(final TransportRequest request) {
    final ClientResponse response =
        new ClientResponse(101, new byte[0], "websocket_handshake", null);
    for (final ClientObserver observer : observers) {
      observer.onResponse(request, response);
    }
  }

  private void notifyFailure(final TransportRequest request, final Throwable error) {
    for (final ClientObserver observer : observers) {
      observer.onFailure(request, error);
    }
  }

  private static final class Adapter implements TransportWebSocket.Listener {
    private final ToriiWebSocketListener delegate;
    private final ToriiWebSocketSessionImpl session;

    Adapter(
        final ToriiWebSocketListener delegate, final ToriiWebSocketSessionImpl session) {
      this.delegate = delegate;
      this.session = session;
    }

    @Override
    public void onOpen(final TransportWebSocket socket) {
      delegate.onOpen(session);
    }

    @Override
    public void onText(
        final TransportWebSocket socket, final CharSequence data, final boolean last) {
      delegate.onText(session, data, last);
    }

    @Override
    public void onBinary(
        final TransportWebSocket socket, final java.nio.ByteBuffer data, final boolean last) {
      delegate.onBinary(session, data, last);
    }

    @Override
    public void onPing(final TransportWebSocket socket, final java.nio.ByteBuffer message) {
      delegate.onPing(session, message);
    }

    @Override
    public void onPong(final TransportWebSocket socket, final java.nio.ByteBuffer message) {
      delegate.onPong(session, message);
    }

    @Override
    public void onError(final TransportWebSocket socket, final Throwable error) {
      delegate.onError(session, error);
      session.fail(error);
    }

    @Override
    public void onClose(
        final TransportWebSocket socket, final int statusCode, final String reason) {
      delegate.onClose(session, statusCode, reason);
      session.close(statusCode, reason);
    }
  }

  private static final class ToriiWebSocketSessionImpl implements ToriiWebSocketSession {
    private final CompletableFuture<TransportWebSocket> ready = new CompletableFuture<>();

    void attach(final TransportWebSocket socket) {
      ready.complete(socket);
    }

    void fail(final Throwable error) {
      ready.completeExceptionally(error);
    }

    @Override
    public CompletableFuture<Void> sendText(final CharSequence data, final boolean last) {
      return ready.thenCompose(ws -> ws.sendText(data, last));
    }

    @Override
    public CompletableFuture<Void> sendBinary(final java.nio.ByteBuffer data, final boolean last) {
      return ready.thenCompose(ws -> ws.sendBinary(data, last));
    }

    @Override
    public CompletableFuture<Void> sendPing(final java.nio.ByteBuffer message) {
      return ready.thenCompose(ws -> ws.ping(message));
    }

    @Override
    public CompletableFuture<Void> sendPong(final java.nio.ByteBuffer message) {
      return ready.thenCompose(ws -> ws.pong(message));
    }

    @Override
    public CompletableFuture<Void> close(final int statusCode, final String reason) {
      return ready.thenCompose(
          ws -> {
            if (!ws.isOpen()) {
              return CompletableFuture.completedFuture(null);
            }
            return ws.close(statusCode, reason);
          });
    }

    @Override
    public boolean isOpen() {
      final TransportWebSocket socket = ready.getNow(null);
      return socket != null && socket.isOpen();
    }

    @Override
    public String subprotocol() {
      final TransportWebSocket socket = ready.getNow(null);
      return socket == null ? "" : socket.subprotocol();
    }
  }

  public static final class Builder {
    private URI baseUri = URI.create("http://localhost:8080");
    private final Map<String, String> defaultHeaders = new LinkedHashMap<>();
    private final List<ClientObserver> observers = new ArrayList<>();
    private WebSocketConnector connector = null;

    public Builder setBaseUri(final URI baseUri) {
      this.baseUri = Objects.requireNonNull(baseUri, "baseUri");
      return this;
    }

    public Builder putDefaultHeader(final String name, final String value) {
      defaultHeaders.put(
          Objects.requireNonNull(name, "name"), Objects.requireNonNull(value, "value"));
      return this;
    }

    public Builder defaultHeaders(final Map<String, String> headers) {
      defaultHeaders.clear();
      if (headers != null) {
        headers.forEach(this::putDefaultHeader);
      }
      return this;
    }

    public Builder addObserver(final ClientObserver observer) {
      observers.add(Objects.requireNonNull(observer, "observer"));
      return this;
    }

    public Builder observers(final List<ClientObserver> values) {
      observers.clear();
      if (values != null) {
        values.forEach(this::addObserver);
      }
      return this;
    }

    public Builder setWebSocketConnector(final WebSocketConnector connector) {
      this.connector = Objects.requireNonNull(connector, "connector");
      return this;
    }

    public ToriiWebSocketClient build() {
      return new ToriiWebSocketClient(this);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/websocket/ToriiWebSocketClient.kt
package org.hyperledger.iroha.sdk.client.websocket

import java.net.URI
import java.net.URLEncoder
import java.nio.ByteBuffer
import java.util.LinkedHashMap
import java.util.concurrent.CancellationException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicBoolean
import org.hyperledger.iroha.sdk.client.ClientObserver
import org.hyperledger.iroha.sdk.client.ClientResponse
import org.hyperledger.iroha.sdk.client.TransportSecurity
import org.hyperledger.iroha.sdk.client.transport.TransportRequest
import org.hyperledger.iroha.sdk.client.transport.TransportWebSocket

/**
 * WebSocket client built on the transport abstractions shared between JVM and Android targets.
 *
 * Inject a concrete WebSocket backend or an explicit application connector. The immutable request
 * contains the complete handshake inputs. A session is attached before its onOpen callback;
 * sends attempted before that callback fail immediately instead of retaining caller buffers.
 */
class ToriiWebSocketClient private constructor(builder: Builder) {

    /** Opens the final immutable handshake request; implementations must emit onOpen once ready. */
    fun interface WebSocketConnector {
        fun connect(
            request: TransportRequest,
            listener: TransportWebSocket.Listener
        ): CompletableFuture<TransportWebSocket>
    }

    private val baseUri: URI = builder.baseUri
    private val defaultHeaders: Map<String, String> = builder.defaultHeaders.toMap()
    private val observers: List<ClientObserver> = builder.observers.toList()
    private val connector: WebSocketConnector = requireNotNull(builder.connector) {
        "WebSocket connector is required; call setWebSocketConnector(...)"
    }

    /** Opens a WebSocket session against the given path. */
    fun connect(path: String, options: ToriiWebSocketOptions?, listener: ToriiWebSocketListener): ToriiWebSocketSession {
        val resolved = options ?: ToriiWebSocketOptions.defaultOptions()
        val request = buildRequest(path, resolved)
        notifyRequest(request)
        val session = ToriiWebSocketSessionImpl()
        val adapter = Adapter(listener, session, request)
        val connection = try {
            connector.connect(request, adapter)
        } catch (failure: Exception) {
            adapter.fail(failure)
            return session
        }
        session.bind(connection)
        connection.whenComplete { socket, error ->
            if (error != null) adapter.fail(error) else session.attach(socket)
        }
        return session
    }

    private fun buildRequest(path: String, options: ToriiWebSocketOptions): TransportRequest {
        val httpUri = appendQueryParameters(resolvePath(path), options.queryParameters)
        val wsUri = toWebSocketUri(httpUri)
        val headers = LinkedHashMap(defaultHeaders)
        options.headers.forEach { (k, v) -> headers[k] = v }
        TransportSecurity.requireWebSocketRequestAllowed(
            "ToriiWebSocketClient",
            baseUri,
            wsUri,
            headers,
        )
        val builder = TransportRequest.builder()
            .setMethod("GET")
            .setUri(wsUri)
            .setTimeout(options.connectTimeout)
        headers.forEach { (k, v) -> builder.addHeader(k, v) }
        if (options.subprotocols.isNotEmpty()) {
            builder.addHeader("Sec-WebSocket-Protocol", options.subprotocols.joinToString(","))
        }
        return builder.build()
    }

    private fun resolvePath(path: String?): URI {
        if (path.isNullOrBlank()) return baseUri
        if (path.startsWith("http://") || path.startsWith("https://") || path.startsWith("ws://") || path.startsWith("wss://")) {
            return URI.create(path)
        }
        val normalized = if (path.startsWith("/")) path.substring(1) else path
        val base = baseUri.toString()
        val joined = if (base.endsWith("/")) base + normalized else "$base/$normalized"
        return URI.create(joined)
    }

    private fun notifyRequest(request: TransportRequest) {
        for (observer in observers) observer.onRequest(request)
    }

    private fun notifyResponse(request: TransportRequest) {
        val response = ClientResponse(101, ByteArray(0), "websocket_handshake", null)
        for (observer in observers) observer.onResponse(request, response)
    }

    private fun notifyFailure(request: TransportRequest, error: Throwable) {
        for (observer in observers) observer.onFailure(request, error)
    }

    private inner class Adapter(
        private val delegate: ToriiWebSocketListener,
        private val session: ToriiWebSocketSessionImpl,
        private val request: TransportRequest,
    ) : TransportWebSocket.Listener {
        private val opened = AtomicBoolean(false)

        override fun onOpen(socket: TransportWebSocket) {
            if (session.attach(socket) && opened.compareAndSet(false, true)) {
                notifyResponse(request)
                delegate.onOpen(session)
            }
        }
        override fun onText(socket: TransportWebSocket, data: String) { delegate.onText(session, data) }
        override fun onBinary(socket: TransportWebSocket, data: ByteBuffer) { delegate.onBinary(session, data) }
        override fun onError(socket: TransportWebSocket, error: Throwable) { fail(error) }
        override fun onClose(socket: TransportWebSocket, statusCode: Int, reason: String) {
            if (session.finish()) delegate.onClose(session, statusCode, reason)
        }
        fun fail(error: Throwable) {
            if (session.finish(error)) {
                notifyFailure(request, error)
                delegate.onError(session, error)
            }
        }
    }

    private class ToriiWebSocketSessionImpl : ToriiWebSocketSession {
        private val lock = Any()
        private var socket: TransportWebSocket? = null
        private var connection: CompletableFuture<TransportWebSocket>? = null
        private var finished = false
        private var failure: Throwable? = null

        fun bind(value: CompletableFuture<TransportWebSocket>) {
            val cancelled = synchronized(lock) { connection = value; finished }
            if (cancelled) value.cancel(false)
        }

        fun attach(value: TransportWebSocket): Boolean {
            val accepted = synchronized(lock) {
                if (finished) false else { socket = value; true }
            }
            if (!accepted) value.close(1000, "session closed")
            return accepted
        }

        fun finish(error: Throwable? = null): Boolean = synchronized(lock) {
            if (finished) false else { finished = true; failure = error; true }
        }

        private fun send(action: (TransportWebSocket) -> CompletableFuture<Void>): CompletableFuture<Void> {
            val target = synchronized(lock) {
                if (finished || socket == null) {
                    return CompletableFuture<Void>().also {
                        it.completeExceptionally(failure ?: IllegalStateException("WebSocket session is not open"))
                    }
                }
                requireNotNull(socket)
            }
            return action(target)
        }

        override fun sendText(data: String): CompletableFuture<Void> = send { it.sendText(data) }
        override fun sendBinary(data: ByteBuffer): CompletableFuture<Void> = send { it.sendBinary(data) }
        override fun close(statusCode: Int, reason: String): CompletableFuture<Void> {
            val target: TransportWebSocket?
            val pending: CompletableFuture<TransportWebSocket>?
            synchronized(lock) {
                if (finished) return CompletableFuture.completedFuture(null)
                target = socket
                pending = connection
                if (target == null) {
                    finished = true
                    failure = CancellationException("WebSocket session closed before handshake")
                }
            }
            if (target != null) return target.close(statusCode, reason)
            pending?.cancel(false)
            return CompletableFuture.completedFuture(null)
        }
        override val isOpen: Boolean get() = synchronized(lock) { !finished && socket?.isOpen() == true }
        override val subprotocol: String get() = synchronized(lock) { if (finished) "" else socket?.subprotocol().orEmpty() }
    }

    class Builder {
        internal var baseUri: URI = URI.create("http://localhost:8080")
        internal val defaultHeaders = LinkedHashMap<String, String>()
        internal val observers = ArrayList<ClientObserver>()
        internal var connector: WebSocketConnector? = null

        fun setBaseUri(baseUri: URI): Builder { this.baseUri = baseUri; return this }
        fun putDefaultHeader(name: String, value: String): Builder { defaultHeaders[name] = value; return this }
        fun defaultHeaders(headers: Map<String, String>?): Builder { defaultHeaders.clear(); headers?.forEach { (k, v) -> putDefaultHeader(k, v) }; return this }
        fun addObserver(observer: ClientObserver): Builder { observers.add(observer); return this }
        fun observers(values: List<ClientObserver>?): Builder { observers.clear(); values?.forEach { addObserver(it) }; return this }
        fun setWebSocketConnector(connector: WebSocketConnector): Builder { this.connector = connector; return this }
        fun build(): ToriiWebSocketClient = ToriiWebSocketClient(this)
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()

        private fun appendQueryParameters(target: URI, params: Map<String, String>): URI {
            if (params.isEmpty()) return target
            val targetText = target.toString()
            val fragmentIndex = targetText.indexOf('#').let { if (it >= 0) it else targetText.length }
            val builder = StringBuilder(targetText.length + 1)
                .append(targetText, 0, fragmentIndex)
            val query = encodeQuery(params)
            if (target.query == null || target.query.isEmpty()) {
                builder.append(if (builder.indexOf("?") >= 0) "&" else "?")
            } else {
                builder.append("&")
            }
            builder.append(query)
            builder.append(targetText, fragmentIndex, targetText.length)
            return URI.create(builder.toString())
        }

        private fun encodeQuery(params: Map<String, String>): String =
            params.entries.joinToString("&") { (k, v) ->
                "${URLEncoder.encode(k, "UTF-8")}=${URLEncoder.encode(v, "UTF-8")}"
            }

        private fun toWebSocketUri(httpUri: URI): URI {
            val scheme = if ("https".equals(httpUri.scheme, ignoreCase = true) || "wss".equals(httpUri.scheme, ignoreCase = true)) "wss" else "ws"
            return URI.create(
                "$scheme://${httpUri.authority}${httpUri.rawPath ?: ""}${if (httpUri.rawQuery != null) "?${httpUri.rawQuery}" else ""}"
            )
        }
    }
}
